src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs

```diff
@@ -0,0 +1,185 @@
+            }
+            if ((data[0] != 0x1) || (data[1] != 0x00))
+            {
+                throw new InvalidOperationException(SR.Format(SR.InvalidOperation_InvalidProlog, ctor.DeclaringType));
+            }
+
+            ParameterInfo[] pi = ctor.GetParameters();
+            info._ctor = ctor;
+            info._ctorArgs = new object?[pi.Length];
+            for (int i = 0; i < pi.Length; ++i)
+            {
+                info._ctorArgs[i] = DecodeCustomAttributeValue(pi[i].ParameterType, data, pos, out pos);
+            }
+            int numNamed = data[pos] + (data[pos + 1] * 256);
+            pos += 2;
+
+            info._namedParamNames = new string[numNamed];
+            info._namedParamValues = new object[numNamed];
+            for (int i = 0; i < numNamed; ++i)
+            {
+                int namedType = data[pos++];
+                int dataType = data[pos++];
+                string? enumTypeName = null;
+
+                if (dataType == 0x55)
+                {
+                    int len2 = DecodeLen(data, pos, out pos);
+                    enumTypeName = StringFromBytes(data, pos, len2);
+                    pos += len2;
+                }
+
+                int len = DecodeLen(data, pos, out pos);
+                string name = StringFromBytes(data, pos, len);
+                info._namedParamNames[i] = name;
+                pos += len;
+
+                if (namedType == 0x53)
+                {
+                    /* Field */
+                    FieldInfo? fi = ctor.DeclaringType!.GetField(name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+                    if (fi == null)
+                    {
+                        throw new InvalidOperationException(SR.Format(SR.InvalidOperation_EmptyFieldForCustomAttribute, ctor.DeclaringType, name));
+                    }
+                    object? val = DecodeCustomAttributeValue(fi.FieldType, data, pos, out pos);
+
+                    if (enumTypeName != null)
+                    {
+                        Type enumType = Type.GetType(enumTypeName)!;
+                        val = Enum.ToObject(enumType, val!);
+                    }
+
+                    info._namedParamValues[i] = val;
+                }
+                else
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_UnknownNamedType, ctor.DeclaringType, namedType));
+                }
+            }
+
+            return info;
+        }
+
+        private static string StringFromBytes(ReadOnlySpan<byte> data, int pos, int len)
+        {
+            return Text.Encoding.UTF8.GetString(data.Slice(pos, len));
+        }
+
+        private static int DecodeLen(ReadOnlySpan<byte> data, int pos, out int rpos)
+        {
+            int len;
+            if ((data[pos] & 0x80) == 0)
+            {
+                len = (int)(data[pos++] & 0x7f);
```

Comment:
nit: perhaps use constants for these masks? (for readability, especially if used in more than one location)